src/dictionary.c

```diff
@@ -22,66 +23,105 @@ int32_t ceil2(int32_t x) {
 
 // Dictonary object ------------------------------------------------------------
 
-static struct dictionary* new_dictionary_impl(SEXP x, bool partial);
+static struct dictionary* new_dictionary_opts(SEXP x, struct dictionary_opts* opts);
 
 
-static int nil_equal(const void* x, R_len_t i, const void* y, R_len_t j) {
+static int nil_p_equal(const void* x, R_len_t i, const void* y, R_len_t j) {
   Rf_error("Internal error: Shouldn't compare NULL in dictionary.");
 }
-static int lgl_equal(const void* x, R_len_t i, const void* y, R_len_t j) {
+static int nil_p_equal_missing(const void* x, R_len_t i) {
+  Rf_error("Internal error: Shouldn't compare NULL in dictionary.");
+}
+
+static int lgl_p_equal_na_equal(const void* x, R_len_t i, const void* y, R_len_t j) {
```

Comment:
Is this supposed to be `lgl_p_equal_na_equal` -> `lgl_p_equal` to match the others?